### ISSUE TITLE
Measles model is too slow

### ISSUE THREAD

**Issue by AJAbkemeier** [pypomp/pypomp#31]
The measles model currently runs too slow. This is due in part to it calling random number generators 3 times for each `rproc` call, but it's slower than the R `pomp` version of the measles model despite `pypomp` running other examined models faster. We should investigate why exactly this is and see how we can improve its performance.

**Comment by ionides** [comment]
@megumi-ybb mentioned that he has previously encountered the same problem of `jax.random.multinomial()` being very slow, but did not find a fix. @hetankevin has used `tfp.distributions.Multinomial(n, logits=logits).sample()` in the [haiti model](https://github.com/hetankevin/diffPomp/blob/main/haiti/pomps.py).

The numbers @ajabkemeier reported today was an equivalent R CPU (ERC) value of 12 for measles on pypomp GPU (i.e., the GPU is faster than a single CPU only by a factor of 12, for greatlakes GPU and CPU services, with vmap applied outside a replication loop). dacca was getting an ERC of 200-400 with vmap, and 50 without. So, the measles multinom here is not just slow but an order of magnitude off.

One possibility is that we're pushing jax too far from its intended use case. `jax.random.multinomial()` is designed for parallelization by broadcasting over the `shape`, but here we are making separate calls for each particle. Ideally, that would not be a problem, but perhaps in practice for a complicated function is it a problem.

Is there any fundamental reason why multinomial resampling (really, binomial resampling since the multinomial is just a loop around binomial) should be fundamentally extremely slow in jax? I guess not.

**Comment by kunyanghe21** [comment]
After spending last night reviewing the JAX official documentation, I have three small suggestions:

In jax.random.multinomial, you can set k=3, so the tiny per-category scan (K=3) is fully unrolled—this removes loop overhead without altering the distribution.

Avoid unnecessary RNG calls (e.g., skip calling multinomial when the population is zero).

Switch to int32, since GPUs run more slowly with int64.

I’ll continue to follow up on this issue and keep working on ways to improve the measles model speed. Here is the toy code:

# Ensure integer counts for RNG (GPU prefers int32)
populations = jnp.asarray(jnp.array([S, E, I]), dtype=jnp.int32)

# Keep probs in float32
rt_final = rt_final.astype(jnp.float32)

def _sample_row(k, n, p_row):
    # if n==0, avoid calling RNG at all
    def do_sample(_):
        return jax.random.multinomial(k, n, p_row, unroll=3)  # fully unroll K=3
    def do_zero(_):
        return jnp.array([0, 0, n], dtype=jnp.int32)          # [to1, to2, stay]
    return jax.lax.cond(n > 0, do_sample, do_zero, operand=None)

kS, kE, kI = jax.random.split(keys[2], 3)
tr_S = _sample_row(kS, populations[0], rt_final[0])
tr_E = _sample_row(kE, populations[1], rt_final[1])
tr_I = _sample_row(kI, populations[2], rt_final[2])
transitions = jnp.stack([tr_S, tr_E, tr_I], axis=0)

**Comment by AJAbkemeier** [comment]
Another thought:

Maybe you could sample variables using an accept-reject algorithm, but you cap the number of rejects at some fixed amount $m$. If you reject $m$ times, you accept the next proposal no matter what. I think the resulting draw would come from

$p_{\text{final}}(x) = (1 - w) \cdot p(x) + w \cdot q(x)$

where $p$ is the target distribution, $q$ is the proposal distribution, and $w$ is the probability of hitting the cap. This seems like a pretty general approach to handling accept-reject algorithms. The Marsaglia and Tsang algorithm for sampling from a gamma distribution appears to have an acceptance probability of at least 0.95 for all choices of $\alpha$, so capping the number of rejections at 0 or 1 shouldn't sacrifice much accuracy.

**Comment by kunyanghe21** [comment]
I also thought about approximate approaches yesterday, but an important point is that we still need to ensure accuracy. I just tested Aaron’s latest code, and the speed improvement is quite significant. In the measles model, version 0.2.2 (previously log-likelihood = −3830.86, SE = 0.38) has been updated to a new version with log-likelihood = −3833.84, SE = 0.56. For this test, I used Np = 5000 and rep = 4. The error is within 3 units, which seems acceptable, though I’m not sure how Ed and others will view it. I think we need to be cautious when choosing approximate methods.

**Comment by AJAbkemeier** [comment]
I imagine one could do something like this:

When fitting the model, used rejection-capped versions of rejection samplers. When you want a casual estimate of the log-likelihood, continue to use the rejection-capped versions of rejection samplers. Once it's time to produce a serious log-likelihood estimate, such as for a paper, you use the true rejection samplers. This might result in a small log-likelihood hit because the model you optimized isn't exactly the same as the final model you used to estimate the log-likelihood, but the model would be a log-likelihood estimate for the exact model you're interested in.

**Comment by jeswheel** [comment]
The approximation concern that @kunyanghe21 raises is an important issue, though I personally lean towards the latter interpretation that @AJAbkemeier hinted at, which is that the real model we use for inference is the thing that we simulate from, and what we write down on paper can be thought of as an approximation of our true model (which is implemented in code, since that is what we use for inference). 

In this regard, I would favor the "fast" method, even if it is approximate, and I think many users would be convinced of this as well. However, from a development standpoint, I would prefer a name like `multinomial_approx` (or equivalent) rather than `fast_multinomial()`, to emphasize that we're doing an approximation, though I'll defer to your preference here.

> Once it's time to produce a serious log-likelihood estimate, such as for a paper, you use the true rejection samplers.

In the past when we fit a model using approximate densities, we are explicit about this fact when describing the model. For instance, we use a normal approximation for the measles measurement model, and when describing the model in the paper, we don't state that we are using a negative-binomial, but instead are explicit that it's a normal approximation to the same distribution. 

I prefer this approach rather than stating we have a particular model, fit a fast approximation to the model, and then evaluating the log-likelihood of the resulting estimate using the exact model---This might not be what you were suggesting, but my thinking here is that if we maximize using an approximation, then evaluate using the "exact" model, then our evaluation of the likelihood at our provided estimate is accurate, but it may not correspond to the maximum. A small difference between the maximum of the approximate and the maximum of the exact model may be small, but if we want to do likelihood-based inference the consequences may be non-negligible. Instead, I see no problem stating that the "exact" model involves an approximate to a common distribution.

**Comment by ionides** [comment]
It's great to have a speed fix. Now the issue is fairly clear, it remains to decide on the solution. Here are some comments.

1. Approximation for the sake of speed is legitimate, but needs validation. In this case, that can be simple: as long as the essentially exact binomial is left as an option, users can check whether or not the approximation is satisfactory in their specific situation by running a few comparisons against the slower but more accurate algorithm.

2. Bounded rejection sampling seems like a good compromise to me. A normal approximation to binomial could be quite bad when np < 1, which may not be uncommon.

3. A curious thing is that an expensive `rprocess` in `pfilter` should favor GPU computation relative to CPU, since the call to `rprocess` (unlike resampling) trivially parallelizes. The random number generation would have to be deeply problematic to be much slower even when the random number generation process is computed separately on each GPU core. We understand very little about the sharding, not least since a goal of jax is to conceal such details unless you really want to get into them.

4. Incidentally, I wonder whether the extended data `fori` approach is inefficient here. If one had a `for` loop then the algorithm would not have to stop for the slowest particle at each Euler timestep. To current understanding, it seems we may have to roll back the extended `fori` for other reasons. Maybe that, together with a bound to prevent large numbers of rejections, maybe enough?

**Comment by AJAbkemeier** [comment]
Capping the number of rejections at 2 for the Marsaglia and Tsang gamma sampling algorithm appears to work well. Running `pfilter()` at the short course MLE with all of the approximate samplers discussed so far yielded an MLE of -3833.76, and the run level 4 quant test runs in 16 minutes and 37 seconds.

I think the next step is to get better approximations for the Poisson and multinomial distributions. The capped rejection method seems to work well enough for the Poisson transformed rejection method, but it wouldn't work well for Knuth's algorithm, and I've obtained pretty poor results when using it with BTRS in the multinomial sampler (either very slow or very inaccurate).

**Comment by AJAbkemeier** [comment]
An update:

I eventually concluded that my existing method of approximating the distributions was too slow and too inaccurate. I ended up looking for other solutions, and I landed on the following 3 papers:

> Giles, M. B. (2016). Algorithm 955: Approximation of the Inverse Poisson Cumulative Distribution Function. ACM Transactions on Mathematical Software, 42(1), 1–22. https://dl.acm.org/doi/10.1145/2699466
> 
> Giles, M. B. & Beentjes, C. (2024). Approximation of an Inverse of the Incomplete Beta Function. In K. Buzzard et al. (Eds.), Mathematical Software – ICMS 2024 (pp. 207–214). Springer Nature Switzerland. https://link.springer.com/10.1007/978-3-031-64529-7_22
> 
> Temme, N. M. Asymptotic Inversion of Incomplete Gamma Functions.

These discuss approximations to the inverse CDFs for the Poisson, binomial, and gamma distributions, respectively. The first includes an algorithm that uses the inverse CDF to speedily sample from the distribution. I added a modified version of this algorithm to pypomp, `rpoisson` (name subject to change). I then added the inverse CDF algorithms from the other two papers and added additional code to handle parameter choices that the approximations weren't great for. These functions are called `rbinomial` and `rgamma`. Based on some tests I ran and some plots, they seem to be pretty accurate. The following plots compare these new samplers to the corresponding samplers in JAX, sampling 100k times from each.

<img width="1224" height="800" alt="Image" src="https://github.com/user-attachments/assets/8f7a370b-c45a-4c92-8bba-922027ae65e7" />

<img width="1224" height="800" alt="Image" src="https://github.com/user-attachments/assets/0157ec7e-2de8-44c8-8b98-bed2f56d1698" />

<img width="1224" height="800" alt="Image" src="https://github.com/user-attachments/assets/76acf408-568e-4e94-b4db-164fbf0f1b83" />

<img width="1224" height="800" alt="Image" src="https://github.com/user-attachments/assets/fa7b91e5-b8e1-4c7a-ac89-f11928e1e6ab" />

There are some things to iron out, but they're much more accurate than the previous approximate samplers I used. 

I tried running them with the measles model on the He10 data set with the following algorithmic parameters:

IF2:
- Particles: 5000
- Iterations: 100
- Replications: 36

pfilter:
- Particles: 5000
- Replications: 36 per IF2 replication

I compared them to an R pomp fit that used the same algorithmic parameters for IF2. The LL is fairly close. I happened to use weaker algorithmic parameters for `pfilter`, so that might partly explain the mild shortfall aside from luck. 

| **City**          | **LL (R)**    | **LL (py)**       | **Difference (py - R)** |
| ----------------- | ------------- | ----------------- | ----------------------- |
| Bedwellty         | -1126.11      | -1125.931049      | 0.1790                  |
| Birmingham        | -3237.48      | -3235.248338      | 2.2317                  |
| Bradford          | -2578.76      | -2580.359582      | -1.5996                 |
| Bristol           | -2682.74      | -2679.580239      | 3.1598                  |
| Cardiff           | -2367.79      | -2367.985208      | -0.1952                 |
| Consett           | -1357.85      | -1359.075571      | -1.2256                 |
| Dalton.in.Furness | -725.26       | -726.393137       | -1.1331                 |
| Halesworth        | -315.61       | -315.790503       | -0.1805                 |
| Hastings          | -1582.94      | -1584.074655      | -1.1347                 |
| Hull              | -2706.94      | -2708.714640      | -1.7746                 |
| Leeds             | -2916.89      | -2916.063772      | 0.8262                  |
| Lees              | -547.39       | -547.171442       | 0.2186                  |
| Liverpool         | -3402.77      | -3403.607585      | -0.8376                 |
| London            | -3798.69      | -3800.850169      | -2.1602                 |
| Manchester        | -3234.94      | -3237.246796      | -2.3068                 |
| Mold              | -296.75       | -296.258421       | 0.4916                  |
| Northwich         | -1189.57      | -1190.094224      | -0.5242                 |
| Nottingham        | -2700.82      | -2701.016104      | -0.1961                 |
| Oswestry          | -695.58       | -695.886029       | -0.3060                 |
| Sheffield         | -2788.92      | -2786.567833      | 2.3522                  |
| **SUM**           | **-40253.80** | **-40257.915297** | **-4.1153**             |

It took 2 hours and 1 minute to run this. In comparison, it would take around 16 hours and to run this in R pomp. It took an average of 6 minutes to fit and evaluate one model. 

Depending on what fits I'm comparing, the ERC for the measles model is now something like 250-300. Adjusting for the cost of using a GPU on Great Lakes compared to a CPU, the ERC is 750-900. Running the above fit and evaluation for the 1422 districts in the full measles data set would take about 2 days and 8 hours using 3 GPUs, costing around $24.
